### PR TITLE
Split gradtape and keras tests

### PR DESCRIPTION
**Pull request by vandanavk**
### Description of changes:
Executing gradtape and keras fit() tests led to a slowdown indicated by `WARNING:tensorflow:Method (on_train_batch_end) is slow compared to the batch update (0.145641). Check your callbacks.`

This occurred only on c4.8x, c5.18x instances.

After multiple experiments, we are splitting the gradienttape and fit() tests into 2 separate files as it seems that there is some state from the GradientTape training that is not getting cleared before fit() training starts. This slowdown in callback was observed even when no smdebug changes were made to the training script.

Below results are for a run of 5 epochs per test

Test results (tests executed in TF 2.1 WIP DLC on c4.8xlarge instance):

`time timeout 600s python tests/zero_code_change_tensorflow2_gradtape_integration_tests.py`
real    2m32.678s
user    3m14.783s
sys     0m9.669s

`time timeout 600s python mock/tests/zero_code_change/tensorflow2_integration_tests.py`
real    6m41.652s
user    50m31.547s
sys     162m40.373s

`time timeout 600s python mock/tests/zero_code_change/tensorflow2_gradtape_integration_tests.py --script-mode`
real    3m6.795s
user    3m31.127s
sys     0m29.912s

`time timeout 600s python mock/tests/zero_code_change/tensorflow2_integration_tests.py --script-mode`
real    7m4.844s
user    53m52.965s
sys     173m53.885s

Tests executed in TF 2.1 WIP GPU DLC on p3.8xlarge:

If the following tests are the first to be executed in a container, they show the following times:
`time timeout 600s python mock/tests/zero_code_change/tensorflow2_gradtape_integration_tests.py`
real    9m30.022s
user    9m48.376s
sys     0m8.984s

`time timeout 600s python mock/tests/zero_code_change/tensorflow2_integration_tests.py`
real    8m27.949s
user    37m41.880s
sys     165m20.208s

Subsequent runs show the following numbers:
`time timeout 600s python mock/tests/zero_code_change/tensorflow2_gradtape_integration_tests.py`
real    4m8.980s
user    4m26.708s
sys     0m9.328s

`time timeout 600s python mock/tests/zero_code_change/tensorflow2_gradtape_integration_tests.py --script-mode`
real    4m4.651s
user    4m40.912s
sys     3m57.780s

`time timeout 600s python mock/tests/zero_code_change/tensorflow2_integration_tests.py --script-mode`
real    8m24.723s
user    39m14.620s
sys     175m2.184s

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/zero_code_change/tensorflow2_gradtape_integration_tests.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        (x_train, y_train), _ = get_keras_data()
+        dataset = tf.data.Dataset.from_tensor_slices(
+            (tf.cast(x_train[..., tf.newaxis] / 255, tf.float32), tf.cast(y_train, tf.int64))
+        )
+        dataset = dataset.shuffle(1000).batch(64)
+
+        opt = tf.keras.optimizers.RMSprop()
+        cce = tf.keras.losses.CategoricalCrossentropy(from_logits=True)
+        train_acc_metric = tf.keras.metrics.SparseCategoricalAccuracy()
+        n_epochs = 5
```
Comment (NihalHarish) [inline tests/zero_code_change/tensorflow2_gradtape_integration_tests.py:50]: Can we reduce the number of epochs?

File: tests/zero_code_change/tensorflow2_gradtape_integration_tests.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+import argparse
+
+# Third Party
+import tensorflow.compat.v2 as tf
+
+# First Party
+import smdebug.tensorflow as smd
+from smdebug.core.utils import SagemakerSimulator
+
+
```
Comment (NihalHarish) [inline tests/zero_code_change/tensorflow2_gradtape_integration_tests.py:19]: Disable progress bar tfds progress bar?

Comment (vandanavk) [reply tests/zero_code_change/tensorflow2_gradtape_integration_tests.py:19]: ok, let me try that

File: tests/zero_code_change/tensorflow2_gradtape_integration_tests.py (hunk 1 above)
Comment (vandanavk) [reply tests/zero_code_change/tensorflow2_gradtape_integration_tests.py:50]: to 3 epochs maybe?

File: tests/zero_code_change/tensorflow2_gradtape_integration_tests.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+def test_keras_v2_default(script_mode: bool = False):
+    # Test default ZCC behavior
+    helper_test_keras_v2_gradienttape(script_mode=script_mode)
+
+
+def test_keras_v2_multi_collections(script_mode: bool = False):
+    # Test multiple collections included in hook json
+    json_file_contents = """
```
Comment (NihalHarish) [inline tests/zero_code_change/tensorflow2_gradtape_integration_tests.py:109]: `tests/zero_code_change/utils.py` has a build_json function.
Can it be used here?

File: tests/zero_code_change/tensorflow2_gradtape_integration_tests.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+            {
+                "S3OutputPath": "s3://sagemaker-test",
+                "LocalPath": "/opt/ml/output/tensors",
+                "HookParameters" : {
+                    "save_steps": "0,1,2,3",
+                    "save_all": true
+                }
+            }
+            """
```
Comment (NihalHarish) [inline tests/zero_code_change/tensorflow2_gradtape_integration_tests.py:152]: tests/zero_code_change/utils.py has a build_json function.
Can it be used here?

File: tests/zero_code_change/tensorflow2_gradtape_integration_tests.py (hunk 1 above)
Comment (NihalHarish) [reply tests/zero_code_change/tensorflow2_gradtape_integration_tests.py:50]: How do multiple epochs help the purpose of this test?

Does the result change with the number of epochs?

If not,  can we use just one epoch?

File: tests/zero_code_change/tensorflow2_gradtape_integration_tests.py (hunk 4 above)
Comment (vandanavk) [reply tests/zero_code_change/tensorflow2_gradtape_integration_tests.py:152]: we'll have to modify the sagemaker simulator to make use of this build_json. maybe do it in a separate PR on master?

File: tests/zero_code_change/tensorflow2_gradtape_integration_tests.py (hunk 3 above)
Comment (vandanavk) [reply tests/zero_code_change/tensorflow2_gradtape_integration_tests.py:109]: we'll have to modify the sagemaker simulator to make use of this build_json. maybe do it in a separate PR on master?

File: tests/zero_code_change/tensorflow2_gradtape_integration_tests.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                    hook.record_tensor_value(tensor_name="accuracy", tensor_value=acc)
+                log = "Epoch %d " % (epoch + 1)
+                log += "Accuracy %.4f" % train_acc_metric.result()
+                print(log)
+                train_acc_metric.reset_states()
+            hook = smd.get_hook()
+            assert hook
+            hook.close()
+            # Check that hook created and tensors saved
+            trial = smd.create_trial(path=sim.out_dir)
```
Comment (NihalHarish) [inline tests/zero_code_change/tensorflow2_gradtape_integration_tests.py:77]: Do we clean the `out_dir` get deleted in this test

Comment (NihalHarish) [reply tests/zero_code_change/tensorflow2_gradtape_integration_tests.py:77]: Okay, `SagemakerSimulator` cleans up the files.

File: tests/zero_code_change/tensorflow2_gradtape_integration_tests.py (hunk 2 above)
Comment (vandanavk) [reply tests/zero_code_change/tensorflow2_gradtape_integration_tests.py:19]: reducing number of epochs to 3. On GPU.

`time timeout 600s python mock/tests/zero_code_change/tensorflow2_gradtape_integration_tests.py`
real    2m48.834s
user    2m58.264s
sys     0m9.412s

`time timeout 600s python mock/tests/zero_code_change/tensorflow2_integration_tests.py`
real    6m3.416s
user    23m54.660s
sys     103m16.680s


tfds.disable_progress_bar may not be relevant in these tests as tfds is not used. probably changing verbose level in fit and evaluate will help. should i try that?

File: tests/zero_code_change/tensorflow2_gradtape_integration_tests.py (hunk 5 above)
Comment (vandanavk) [reply tests/zero_code_change/tensorflow2_gradtape_integration_tests.py:77]: sagemaker simulator removes the existing out_dir at the beginning of the next test. 

```
class SagemakerSimulator(object):
    """
    Creates an environment variable pointing to a JSON config file, and creates the config file.
    Used for integration testing with zero-code-change.

    If `disable=True`, then we still create the `out_dir` directory, but ignore the config file.
    """

    def __init__(
        self,
        json_config_path="/tmp/zcc_config.json",
        tensorboard_dir="/tmp/tensorboard",
        training_job_name="sm_job",
        json_file_contents="{}",
        cleanup=True,
    ):
        self.out_dir = DEFAULT_SAGEMAKER_OUTDIR
        self.json_config_path = json_config_path
        self.tb_json_config_path = DEFAULT_SAGEMAKER_TENSORBOARD_PATH
        self.tensorboard_dir = tensorboard_dir
        self.training_job_name = training_job_name
        self.json_file_contents = json_file_contents
        self.cleanup = cleanup

    def __enter__(self):
        if self.cleanup is True:
            shutil.rmtree(self.out_dir, ignore_errors=True)
        
```

File: tests/zero_code_change/tensorflow2_gradtape_integration_tests.py (hunk 1 above)
Comment (vandanavk) [reply tests/zero_code_change/tensorflow2_gradtape_integration_tests.py:50]: result does not change. based on previous discussions, it is recommended to have at least 2 epochs to ensure there are not issues when transitioning between one epoch and another.

**Comment by vandanavk** [comment]
@NihalHarish @Vikas-kum all review comments addressed. once CI passes, I'll change the base branch back to 0.7.2

**Pull request merged**